pertpy/tools/_sccoda.py

```diff
@@ -0,0 +1,2054 @@
+        reference_cell_type: str = "automatic",
+        automatic_reference_absence_threshold: float = 0.05,
+    ) -> AnnData:
+        """Handles data preprocessing, covariate matrix creation, reference selection, and zero count replacement.
+
+        Args:
+            sample_adata (AnnData): anndata object with cell counts as sample_adata.X and covariates saved in sample_adata.obs.
+            formula (str): R-style formula for building the covariate matrix.
+                Categorical covariates are handled automatically, with the covariate value of the first sample being used as the reference category.
+                To set a different level as the base category for a categorical covariate, use "C(<CovariateName>, Treatment('<ReferenceLevelName>'))"
+            reference_cell_type (str, optional): Column name that sets the reference cell type.
+                Reference the name of a column. If "automatic", the cell type with the lowest dispersion in relative abundance that is present in at least 90% of samlpes will be chosen. Defaults to "automatic".
+            automatic_reference_absence_threshold (float, optional): If using reference_cell_type = "automatic", determine the maximum fraction of zero entries for a cell type
+                to be considered as a possible reference cell type. Defaults to 0.05.
+
+        Returns:
+            AnnData
+        """
+        dtype = "float64"
+
+        # Convert count data to float64 (needed for correct inference)
+        sample_adata.X = sample_adata.X.astype(dtype)
+
+        # Build covariate matrix from R-like formula, save in obsm
+        covariate_matrix = pt.dmatrix(formula, sample_adata.obs)
+        covariate_names = covariate_matrix.design_info.column_names[1:]
+        sample_adata.obsm["covariate_matrix"] = np.array(covariate_matrix[:, 1:]).astype(dtype)
+
+        cell_types = sample_adata.var.index.to_list()
+
+        # Invoke instance of the correct model depending on reference cell type
+        # Automatic reference selection (dispersion-based)
+        if reference_cell_type == "automatic":
+            percent_zero = np.sum(sample_adata.X == 0, axis=0) / sample_adata.X.shape[0]
+            nonrare_ct = np.where(percent_zero < automatic_reference_absence_threshold)[0]
+
+            if len(nonrare_ct) == 0:
+                raise ValueError(
+                    "No cell types that have large enough presence! Please increase automatic_reference_absence_threshold"
+                )
+
+            rel_abun = sample_adata.X / np.sum(sample_adata.X, axis=1, keepdims=True)
+
+            # select reference
+            cell_type_disp = np.var(rel_abun, axis=0) / np.mean(rel_abun, axis=0)
+            min_var = np.min(cell_type_disp[nonrare_ct])
+            ref_index = np.where(cell_type_disp == min_var)[0][0]
+
+            ref_cell_type = cell_types[ref_index]
+            print(f"Automatic reference selection! Reference cell type set to {ref_cell_type}")
```

Comment:
```suggestion
            print(f"[bold blue]Automatic reference selection! Reference cell type set to {ref_cell_type}")
```

Also need to `from rich import print` above